irobotclient/main.py

```
@@ -0,0 +1,100 @@
+    """
+
+    :param response:
+    :param output_dir:
+    :return:
+    """
+    try:
+        file_name = (path.split(response.url))[1]
+
+        if response.headers["content-type"] == "application/octet-stream":
+            with open(f"{output_dir}{file_name}", "wb") as file:
+                file.write(bytes(response.content))
+        else:
+            with open(f"{output_dir}{file_name}", "w") as file:
+                file.write(response.content)
+    except OSError as err:
+        _print_error_details(err)
+    except:
+        raise IrobotClientException(errno=errno.ECONNABORTED, message="Cannot write content to file.")
+
+
+def _run():
+    try:
+
+        # Set configurations from command line and/or environment.
+        config_details = configuration_handler.run()
+
+        file_extensions = request_formatter.get_file_extensions(config_details.input_file, config_details.no_index)
+        url = request_formatter.get_url_request_path(config_details.url, config_details.input_file)
+        headers = request_formatter.get_header(config_details.token)
+
+        if not file_extensions:
+            request_handler = Requester(url, headers)
+            response = request_handler.get_data()
+            _download_data(response, config_details.output_dir)
+            # TODO - checksum test if possible
+        else:
+            for ext in file_extensions:
+                # TODO - handle index file issues whereby a bam bai file may not be present but a pbi might.
+                request_handler = Requester(url + ext, headers)
+                response = request_handler.get_data()
+                _download_data(response, config_details.output_dir)
+                # TODO - checksum test for each file if possible
+
+        print("Exiting....")
+        exit()
+    except IrobotClientException as err:
+        _print_error_details(err)
+    except OSError as err:
+        _print_error_details(err)
+    except Exception as err:
```

It would be better to let the excpetion go unhindered as the client will crash with a useful stack trace.

---

@colin-nolan, @Xophmeister  Do you think I should do this for all exceptions then?  Surely the stack trace is always going to be better than whatever error message I can produce.....

---

As with all things engineering: It depends™. Stack traces are useful for debugging, but not particularly friendly for end users. It's a judgement call as to who your users will be and what you think they'll find more useful. Also, some exceptions are obvious and won't necessarily need a stack trace (e.g., file not found).

That only applies to exceptions that you let bubble to the surface, rather than any you catch and handle earlier down the stack. As such, for now, I would err on the side of stack traces for this class of exceptions. In time, you'll learn what "normal behaviour" is; i.e., under normal operating conditions, what exceptions usually bubble to the surface. These will either be bugs or things you can deal with/explain to the user without bombarding them with a stack trace. Then, all that's left will be "weird behaviour", that will need a stack trace to be understood.

---

I have removed the catch all handling of unexpected exceptions for now and will let the program crash with a stack trace.